crates/ruma-client-api/src/discovery/discover_homeserver.rs

```diff
@@ -112,3 +133,187 @@ impl TileServerInfo {
+#[cfg(feature = "unstable-msc4143")]
+#[derive(Clone, Debug, Deserialize, Serialize)]
+#[cfg_attr(not(ruma_unstable_exhaustive_types), non_exhaustive)]
+#[serde(tag = "type")]
+pub enum RtcFocusInfo {
+    /// A LiveKit RTC focus.
+    #[serde(rename = "livekit")]
+    LiveKit(LiveKitRtcFocusInfo),
+
+    /// A custom RTC focus.
+    #[doc(hidden)]
+    #[serde(untagged)]
+    _Custom(CustomRtcFocusInfo),
+}
+
+#[cfg(feature = "unstable-msc4143")]
+impl RtcFocusInfo {
+    /// A constructor to create a custom RTC focus.
+    ///
+    /// Prefer to use the public variants of `RtcFocusInfo` where possible; this constructor is
+    /// meant be used for unsupported focus types only and does not allow setting arbitrary data
+    /// for supported ones.
+    ///
+    /// # Errors
+    ///
+    /// Returns an error if the `focus_type` is known and serialization of `data` to the
+    /// corresponding `RtcFocusInfo` variant fails.
+    pub fn new(focus_type: &str, data: JsonObject) -> serde_json::Result<Self> {
+        fn deserialize_variant<T: DeserializeOwned>(obj: JsonObject) -> serde_json::Result<T> {
+            serde_json::from_value(JsonValue::Object(obj))
+        }
+
+        Ok(match focus_type {
+            "livekit" => Self::LiveKit(deserialize_variant(data)?),
+            _ => Self::_Custom(CustomRtcFocusInfo { focus_type: focus_type.to_owned(), data }),
+        })
+    }
+
+    /// Creates a new `RtcFocusInfo`.
+    pub fn livekit(service_url: String) -> Self {
+        Self::LiveKit(LiveKitRtcFocusInfo { service_url })
+    }
+
+    /// Returns a reference to the focus type of this RTC focus.
+    pub fn focus_type(&self) -> &str {
+        match self {
+            Self::LiveKit(_) => "livekit",
+            Self::_Custom(custom) => &custom.focus_type,
+        }
+    }
+
+    /// Returns the associated data.
+    ///
+    /// The returned JSON object won't contain the `focus_type` field, please use
+    /// [`.focus_type()`][Self::focus_type] to access that.
+    ///
+    /// Prefer to use the public variants of `RtcFocusInfo` where possible; this method is meant to
+    /// be used for custom focus types only.
+    pub fn data(&self) -> Cow<'_, JsonObject> {
+        fn serialize<T: Serialize>(object: &T) -> JsonObject {
+            match serde_json::to_value(object).expect("rtc focus type serialization to succeed") {
+                JsonValue::Object(object) => object,
+                _ => panic!("all rtc focus types must serialize to objects"),
+            }
+        }
+
+        match self {
+            Self::LiveKit(info) => Cow::Owned(serialize(info)),
+            Self::_Custom(info) => Cow::Borrowed(&info.data),
+        }
+    }
+}
+
+/// Information about a LiveKit RTC focus.
+#[cfg(feature = "unstable-msc4143")]
+#[derive(Clone, Debug, Deserialize, Serialize)]
+#[cfg_attr(not(ruma_unstable_exhaustive_types), non_exhaustive)]
+pub struct LiveKitRtcFocusInfo {
+    /// The service URL for the LiveKit service.
```

Comment:
It seems a bit redundant, maybe:

```suggestion
    /// The URL for the LiveKit service.
```